File: backends/php-laravel/app/Support/Http/Responses/AbstractResponse.php
<?php

/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */

declare(strict_types=1);

namespace App\Support\Http\Responses;

use App\Support\Traits\Assign;
use App\Support\Traits\Initialize;
use Illuminate\Support\Str;
use ReflectionClass;

/**
 * 基底Responseクラスです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 * @package App\Support\Http\Responses
 *
 * @method int|null getVersion()
 */
abstract class AbstractResponse implements Response
{
    use Assign;
    use Initialize;

    protected ?int $version = null;

    /**
     * コンストラクタ
     */
    public function __construct()
    {
        $this->initializer();
    }

    /**
     * @inheritdoc}
     */
    #[\Override]
    public function attributes(): array|string
    {
        $result = [];
        foreach ($this->getResponseKeys() as $property) {
            if (array_key_exists($property, $this->getExcludeKeys())) {
                continue;
            }

            $result[Str::snake($property)] = $this->$property;
        }
        return $result;
    }

    /**
     * レスポンス出力キーを取得します。
     *
     * @return string[] レスポンス出力キー配列
     */
    protected function getResponseKeys(): array
    {
        $properties = (new ReflectionClass($this))->getProperties();
        $result = [];
        foreach ($properties as $property) {
            $result[] = $property->getName();
        }

        return $result;
    }

    /**
     * レスポンスから除外するキーを取得します。
     *
     * @return string[] レスポンスから除外するキー配列
     */
    protected function getExcludeKeys(): array
    {
        return [];
    }
}

File: backends/java-spring/src/main/java/com/authorization/support/http/responses/AbstractResponse.java
/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */
package com.authorization.support.http.responses;

import java.lang.reflect.Field;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

/**
 * 基底Responseクラスです（PHP版 App\Support\Http\Responses\AbstractResponse 相当）。
 * 派生クラスのフィールドを1件ずつMapへ書き出す手書きコードを避けるため、
 * リフレクションでフィールド名をsnake_caseに変換して {@link #attributes()} を組み立てます。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
public abstract class AbstractResponse {

    /**
     * 出力から除外するフィールド名を返します。既定では除外しません。
     *
     * @return 除外するフィールド名の集合
     */
    protected Set<String> getExcludeKeys() {
        return Set.of();
    }

    /**
     * 派生クラスの全フィールドをsnake_caseキーのMapへ変換します。
     *
     * @return JSON化用のMap
     */
    public Map<String, Object> attributes() {
        Set<String> excludes = getExcludeKeys();
        Map<String, Object> result = new LinkedHashMap<>();
        for (Field field : getClass().getDeclaredFields()) {
            if (excludes.contains(field.getName())) {
                continue;
            }
            field.setAccessible(true);
            try {
                result.put(toSnakeCase(field.getName()), field.get(this));
            } catch (IllegalAccessException e) {
                throw new IllegalStateException("failed to read response field: " + field.getName(), e);
            }
        }
        return result;
    }

    /**
     * キャメルケースの文字列をsnake_caseへ変換します。
     *
     * @param name キャメルケース文字列
     * @return snake_case文字列
     */
    private static String toSnakeCase(String name) {
        StringBuilder sb = new StringBuilder();
        for (char c : name.toCharArray()) {
            if (Character.isUpperCase(c)) {
                sb.append('_').append(Character.toLowerCase(c));
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }
}
